File: hello_world/helloworld.c
#include <linux/types.h>
#include <linux/module.h>
#include <linux/kernel.h>

static int __init helloworld_init(void)
{
    printk( "Hello, World\n");
    return 0;
}

static void __exit helloworld_exit(void)
{
    printk("Goodbye, exit\n");
}

module_init(helloworld_init);
module_exit(helloworld_exit);


MODULE_LICENSE("GPL");
MODULE_AUTHOR("Your Name");
MODULE_DESCRIPTION("A simple Hello World Kernel Module");

File: hello_world/helloworld.rs
#![no_std]
#![no_main]

use kernel::prelude::*;
use kernel::module;

module! {
    type: HelloWorldModule,
    name: b"helloworld",
    author: b"Your Name",
    description: b"A simple Hello World Kernel Module",
    license: b"GPL",
}

struct HelloWorldModule;

impl kernel::Module for HelloWorldModule {
    fn init() -> Result<Self> {
        pr_info!("Hello, World\n");
        Ok(Self)
    }
}

impl Drop for HelloWorldModule {
    fn drop(&mut self) {
        pr_info!("Goodbye, exit\n");
    }
}
